nextcord/http.py

```diff
@@ -269,152 +624,232 @@ async def request(
+                                    url_rate_limit.bucket,
                                 )
-                                self._global_over.clear()
-                            else:
-                                self._dispatch(
-                                    "http_ratelimit",
-                                    int(limit),
-                                    int(remaining),
-                                    retry_after,
-                                    bucket,
-                                    response.headers.get("X-RateLimit-Scope"),
+                                correct_rate_limit = self._buckets[url_rate_limit.bucket]
+                                self._set_url_rate_limit(
+                                    route.method, route, auth, correct_rate_limit
                                 )
+                                if correct_rate_limit.bucket:
+                                    # Signals to all requests waiting to acquire to migrate.
+                                    url_rate_limit.migrate_to(correct_rate_limit.bucket)
+                                else:
+                                    raise ValueError(
+                                        f"Migrating to bucket {correct_rate_limit.bucket}, but "
+                                        f"correct_rate_limit.bucket is falsey. This is likely an internal Nextcord "
+                                        f"issue and should be reported."
+                                    )
+                                # Update the correct RateLimit object with our findings.
+                                await correct_rate_limit.update(response)
+                            elif url_rate_limit.bucket is not None:
+                                self._buckets[url_rate_limit.bucket] = url_rate_limit
+
+                            # even errors have text involved in them so this is safe to call
+                            ret = await json_or_text(response)
+
+                            if response.status >= 400:
+                                # >= 500 was considered, but stuff like 501 and 505+ are not good to retry on.
+                                if response.status in {500, 502, 504}:
+                                    _log.info(
+                                        "Path %s encountered a Discord server issue, retrying.",
+                                        rate_limit_path,
+                                    )
+                                    await asyncio.sleep(1 + retry_count * 2)
+                                    should_retry = True
+                                elif response.status == 401:
+                                    _log.warning(
+                                        "Path %s resulted in error 401, rejected authorization?",
+                                        rate_limit_path,
+                                    )
+                                    raise HTTPException(response, ret)
```

Comment:
Any plans to make a specific Unauthorized exception class?

Reply:
I was going to say no, but then I remembered that `request()` actually throwing a status `401` related error was never specifically handled, or suppose to happen in the old implementation. Adding and fixing!